### PR TITLE
JS: deprecate TrackedNodes.qll

### PR DESCRIPTION
**Pull request by erik-krogh**
I ended up putting `deprecated` annotations on all the private predicates to silence any warning about using `deprecated` features.

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/dataflow/TrackedNodes.qll [hunk 1]
```diff
@@ -1,4 +1,6 @@
 /**
+ * DEPRECATED: Use `TypeTracking.qll` or a `DataFlow::Configuration` from `Configuration.qll` instead.
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/dataflow/TrackedNodes.qll:2]: Lets avoid encouraging spurious introductions of `Configuration`. The curious programmers will find that solution by themselves. I think we should show an example of how one can address this deprecation. I have started below.

```suggestion
 * DEPRECATED: Use `TypeTracking.qll` instead.
 * 
 * The following `TrackedNode` usage is less expressive than the type tracking usage below.
 *
 * ` ` ` <-- remove spaces
 * class MyTrackedNode extends TrackedNode {
 *    MyTrackedNode() { isInteresting(this) }
 * }
 *
 * DataFlow::Node getMyTrackedNodeLocation(MyTrackedNode n) {
 *   n.flowsTo(result)
 * }
 *
 ...


```

**Review by esbena — CHANGES_REQUESTED** [review]

File: javascript/ql/src/semmle/javascript/dataflow/TrackedNodes.qll [hunk 2]
```diff
@@ -1,6 +1,32 @@
...
+ * }
+ * ```
+ *
+ * ```
+ * DataFlow::Node getMyTrackedNodeLocation(DataFlow::Node start, DataFlow::TypeTracker t) {
+ *   t.start() and
+ *   isInteresting(result) and
+ *   result = start
+ *   or
+ *   exists(DataFlow::TypeTracker t2 | t = t2.smallstep(getMyTrackedNodeLocation(start, t2), result))
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/dataflow/TrackedNodes.qll:22]:
```suggestion
 *   exists(DataFlow::TypeTracker t2 | t = t2.step(getMyTrackedNodeLocation(start, t2), result))
```

File: javascript/ql/src/semmle/javascript/dataflow/TrackedNodes.qll [hunk 3]
```diff
@@ -1,6 +1,32 @@
...
+ * }
+ *
+ * DataFlow::Node getMyTrackedNodeLocation(DataFlow::Node n) {
+ *   result = getMyTrackedNodeLocation(n, DataFlow::TypeTracker::end())
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/dataflow/TrackedNodes.qll:26]:
```suggestion
 *   getMyTrackedNodeLocation(n, DataFlow::TypeTracker::end()).flowsTo(result)
```

**Review by esbena — COMMENTED** [review]
Hmm. Do we really need to mention `smallstep` (see suggestions)?

**Comment by erik-krogh** [comment]
> Hmm. Do we really need to mention `smallstep` (see suggestions)?

I know we prefer the big steps.  
But `TrackedNode`s include small steps, and that is sometimes needed (e.g. when the starting point is a string-constant).  
So using only `SourceNode` steps wouldn't be equivalent.   
That is why I mentioned `smallStep`. 

I could additionally add a big-step example?

**Comment by esbena** [comment]
Right. Could we change the tutorial to list the bigstep example first with a sentence like:

 > The following `TrackedNode` usage usually equivalent type tracking usage below.

And then follow up with:

> In rare cases additional tracking is required, for instance when tracking string constants, the following type tracking formulation is required instead.

**Review by esbena — APPROVED** [review]

**Pull request merged**